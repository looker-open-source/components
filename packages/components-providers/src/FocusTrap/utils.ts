/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

// The following pulls heavily from https://github.com/focus-trap/focus-trap
// with the main differences being:
// 1. Only need the activate/deactivate functions & associated utils â€“
// multiple focus traps are managed with FocusTrapProvider
// 2. No need for most of the configurable features
// 3. The text of an input should not be selected on focus if it is readonly

import { tabbable, isFocusable, FocusableElement } from 'tabbable'

const isSelectableInput = (
  node: FocusableElement
): node is HTMLInputElement => {
  const nodeAsInput = node as HTMLInputElement
  return (
    nodeAsInput.tagName !== undefined &&
    nodeAsInput.tagName.toLowerCase() === 'input' &&
    typeof nodeAsInput.select === 'function' &&
    !nodeAsInput.readOnly
  )
}

const checkFocusLost = () => {
  // If focused landed on body or nothing, return it to previous element
  // otherwise it's probably already on something useful/intentional
  return document.activeElement
    ? document.activeElement.tagName === 'BODY'
    : true
}

export const activateFocusTrap = (container: HTMLElement) => {
  const nodeFocusedBeforeActivation = document.activeElement
  let firstTabbableNode: FocusableElement = container
  let lastTabbableNode: FocusableElement = container
  let mostRecentlyFocusedNode: FocusableElement | null = null

  const getInitialFocusNode = () => {
    let node
    if (container.contains(document.activeElement)) {
      node = document.activeElement as HTMLElement
    } else {
      // Look for data-autofocus b/c React strips autofocus from dom
      // https://github.com/facebook/react/issues/11851
      const autoFocusElement = container.querySelector(
        '[data-autofocus="true"]'
      ) as HTMLElement

      // In the absence of autofocus, the surface will have initial focus
      const surfaceElement = container.querySelector(
        '[data-overlay-surface="true"]'
      ) as HTMLElement
      node = autoFocusElement || surfaceElement || container
    }

    if (!node || !isFocusable(node)) {
      throw new Error(
        'Your focus trap needs to have at least one focusable element'
      )
    }

    return node
  }

  const updateTabbableNodes = () => {
    const tabbableNodes = tabbable(container)
    firstTabbableNode = tabbableNodes[0] || getInitialFocusNode()
    lastTabbableNode =
      tabbableNodes[tabbableNodes.length - 1] || getInitialFocusNode()
  }

  const tryFocus = (node: FocusableElement) => {
    if (node === document.activeElement) return
    if (!node || !node.focus) {
      tryFocus(getInitialFocusNode())
      return
    }
    node.focus()
    mostRecentlyFocusedNode = node
    if (isSelectableInput(node)) {
      node.select()
    }
  }

  // This needs to be done on mousedown and touchstart instead of click
  // so that it precedes the focus event.
  const checkPointerDown = function (e: MouseEvent | TouchEvent) {
    if (!container.contains(e.target as Node)) {
      // immediately deactivate the trap
      deactivate()
    }
  }

  // In case focus escapes the trap for some strange reason, pull it back in.
  const checkFocusIn = (e: FocusEvent) => {
    // In Firefox when you Tab out of an iframe the Document is briefly focused.
    if (container.contains(e.target as Node) || e.target instanceof Document) {
      return
    }
    e.stopImmediatePropagation()
    tryFocus(mostRecentlyFocusedNode || getInitialFocusNode())
  }

  const checkKey = (e: KeyboardEvent) => {
    if (e.key === 'Tab' || e.keyCode === 9) {
      // Hijack Tab events on the first and last focusable nodes of the trap,
      // in order to prevent focus from escaping. If it escapes for even a
      // moment it can end up scrolling the page and causing confusion so we
      // kind of need to capture the action at the keydown phase.
      updateTabbableNodes()
      if (e.shiftKey && e.target === firstTabbableNode) {
        e.preventDefault()
        tryFocus(lastTabbableNode)
        return
      }
      if (!e.shiftKey && e.target === lastTabbableNode) {
        e.preventDefault()
        tryFocus(firstTabbableNode)
      }
    }
  }

  document.addEventListener('focusin', checkFocusIn, true)
  document.addEventListener('mousedown', checkPointerDown, {
    capture: true,
    passive: false,
  })
  document.addEventListener('touchstart', checkPointerDown, {
    capture: true,
    passive: false,
  })
  document.addEventListener('keydown', checkKey, {
    capture: true,
    passive: false,
  })

  const t = setTimeout(() => {
    tryFocus(getInitialFocusNode())
  }, 0)

  // Deactivate function
  const deactivate = () => {
    clearTimeout(t)
    document.removeEventListener('focusin', checkFocusIn, true)
    document.removeEventListener('mousedown', checkPointerDown, true)
    document.removeEventListener('touchstart', checkPointerDown, true)
    document.removeEventListener('keydown', checkKey, true)

    // If focus lands on the body, move it back to where it was before the trap
    if (checkFocusLost() && nodeFocusedBeforeActivation) {
      const elementToFocus = nodeFocusedBeforeActivation as HTMLElement
      // When returning focus to the element focused prior to focus trap activation
      // opening the tooltip at that moment is undesirable
      elementToFocus.setAttribute('data-notooltip', 'true')
      elementToFocus.focus()
      // Clear the data attribute immediately so that tooltip will work the next time
      elementToFocus.removeAttribute('data-notooltip')
    }
  }
  return deactivate
}
