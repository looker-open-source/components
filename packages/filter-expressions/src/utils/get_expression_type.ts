/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { Category } from '@looker/sdk'
import type { IDashboardFilter, ILookmlModelExploreField } from '@looker/sdk'
import type { FilterExpressionType } from '../types/filter_type'

interface FilterExpressionTypeMap {
  [type: string]: FilterExpressionType
}

const getTimeframeExpressionType = (fieldType?: string) => {
  if (!fieldType) return 'date'

  const isDateTime = ['date_time', 'hour', 'minute', 'second'].some(
    (timeString: string) => {
      return fieldType.indexOf(timeString) > -1
    }
  )
  if (isDateTime) return 'date_time'

  return 'date'
}

/**
 * yields a value for the 'expressionType' prop on the Filter component
 */
export const getExpressionTypeFromField = (
  field: ILookmlModelExploreField
): FilterExpressionType => {
  if (field?.category === Category.parameter && field?.type === 'number') {
    // If parameter field has enumerations it always parsed as 'tier' but
    // visually should be treated differently if number type defined in LookML
    // bugs: b/187940941, b/199507872
    return field.type as FilterExpressionType
  }
  if (field.enumerations) {
    return 'tier'
  }
  if (field.is_numeric) {
    return 'number'
  }
  if (field.is_timeframe) {
    return getTimeframeExpressionType(field.type)
  }
  if (field.type === 'location' || field.type === 'location_bin_level') {
    return 'location'
  }
  return 'string'
}

/**
 * Returns a valid filter expression type from a DashboardFilter object
 * depending if the filter is of type field or not
 */
export const getExpressionType = (
  filter: Pick<IDashboardFilter, 'type' | 'field'>
): FilterExpressionType => {
  if (filter.field) {
    return getExpressionTypeFromField(filter.field as ILookmlModelExploreField)
  }

  const filterExpressionMap: FilterExpressionTypeMap = {
    number_filter: 'number',
    string_filter: 'string',
    date_filter: 'date',
    // Catch for broken cases where field is not defined, display as a
    // string filter
    field_filter: 'string',
  }
  return filterExpressionMap[filter.type!]
}
