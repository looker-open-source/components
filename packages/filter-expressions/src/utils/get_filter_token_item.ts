/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { FilterASTNode, FilterModel, FilterExpressionType } from '../types'

import { addDays, dateToFilterDateTimeModel, sanitizeDate } from './date'
import { sanitizeNumber } from './number'
import { sanitizeString } from './string'

// transforms the AST for user with filter tokens
export const getFilterTokenItem = (
  root: FilterASTNode,
  filterType: FilterExpressionType,
  configType: string
): FilterModel => {
  let item = { ...root } as FilterModel
  switch (filterType) {
    case 'number':
      item = getNumberFilterTokenItem(item, configType)
      break
    case 'string':
      item = sanitizeString({ ...item, type: 'match' })
      break
    case 'date':
      item = getDateFilterTokenItem(item, configType)
      break
  }
  return { ...item, is: true, left: null, right: null }
}

/**
 * Transforms the number filter item type according to filter config type
 */
const getNumberFilterTokenItem = (
  item: FilterModel,
  configType: string
): FilterModel => {
  switch (configType) {
    case 'range_slider': {
      const value = (item.value && item.value?.[0]) || 0
      return sanitizeNumber({
        ...item,
        type: 'between',
        low: item.low ?? value,
        high: item.high ?? value,
      })
    }
    case 'slider':
      return sanitizeNumber({
        ...item,
        type: '=',
        value: item.value?.length ? item.value : item.low ? [item.low] : [0],
      })
    default:
      return sanitizeNumber({ ...item, type: '=' })
  }
}

const getDateFilterTokenItem = (
  item: FilterModel,
  configType: string
): FilterModel => {
  switch (configType) {
    case 'day_picker':
      return sanitizeDate({ ...item, type: 'on' })
    case 'day_range_picker': {
      // get default dateItem
      const dateItem =
        item.date || dateToFilterDateTimeModel(new Date(Date.now()))
      const {
        start = dateItem,
        end = dateToFilterDateTimeModel(addDays(new Date(Date.now()), 1)),
      } = item
      return sanitizeDate({ ...item, type: 'range', start, end })
    }
    default:
      return sanitizeDate(item)
  }
}
