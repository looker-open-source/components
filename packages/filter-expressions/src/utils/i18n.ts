/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { Resource } from 'i18next'
import i18next from 'i18next'
import merge from 'lodash/merge'
import _availableLocales from '../locales/available_locales'
import deDE from '../locales/de-DE'
import en from '../locales/en'
import esES from '../locales/es-ES'
import frFR from '../locales/fr-FR'
import itIT from '../locales/it-IT'
import jaJP from '../locales/ja-JP'
import koKR from '../locales/ko-KR'
import nlNL from '../locales/nl-NL'
import plPL from '../locales/pl-PL'
import ptBR from '../locales/pt-BR'
import ptPT from '../locales/pt-PT'
import ruRU from '../locales/ru-RU'
import svSE from '../locales/sv-SE'
import trTR from '../locales/tr-TR'
import zhCN from '../locales/zh-CN'
import zhTW from '../locales/zh-TW'

export const availableLocales = _availableLocales

export const i18nResources = {
  'de-DE': deDE,
  en,
  'es-ES': esES,
  'fr-FR': frFR,
  'it-IT': itIT,
  'ja-JP': jaJP,
  'ko-KR': koKR,
  'nl-NL': nlNL,
  'pl-PL': plPL,
  'pt-BR': ptBR,
  'pt-PT': ptPT,
  'ru-RU': ruRU,
  'sv-SE': svSE,
  'tr-TR': trTR,
  'zh-CN': zhCN,
  'zh-TW': zhTW,
}

export interface I18nState {
  /**
   * A dictionary of namespaces containing key value pairs of translations
   */
  resources: Resource
  /**
   * The current locale value
   */
  locale: string
}

/**
 * Directly initialize the localization instance
 */
export const i18nInit = async ({
  resources,
  locale = 'en',
}: Partial<I18nState> = {}) => {
  await i18next.init({
    nsSeparator: false,
    keySeparator: false,
    fallbackLng: 'en',
    lng: locale,
    interpolation: {
      escapeValue: false,
    },
    // add optional resources from args
    resources: merge(i18nResources, resources),
  })
  return i18next
}
