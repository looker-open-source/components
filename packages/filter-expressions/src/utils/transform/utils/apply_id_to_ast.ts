/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { FilterASTNode } from '../../../types'

/**
 * Traverses the AST and assigns an id attribute to each node
 */
const applyId = (root: FilterASTNode): FilterASTNode => {
  let id = 0
  const traverse = (node: FilterASTNode): FilterASTNode => {
    const { left, right } = node
    let newLeft, newRight
    if (left) {
      newLeft = traverse(left)
    }
    id += 1
    const newNode = { ...node, id }
    if (right) {
      newRight = traverse(right)
    }
    return { ...newNode, left: newLeft, right: newRight }
  }
  return traverse(root)
}

export default applyId
