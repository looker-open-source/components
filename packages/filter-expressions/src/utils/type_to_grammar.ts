/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import {
  dateGrammar,
  locationGrammar,
  numberGrammar,
  stringGrammar,
  tierGrammar,
} from '../grammars'
import type {
  FilterASTNode,
  FilterItemToStringFunction,
  FilterToStringFunctionType,
  FilterExpressionType,
  TransformFunction,
} from '../types'
import {
  anyNumberItem,
  anyStringItem,
  anyValue,
  dateFilterTypes,
  locationFilterTypes,
  numberFilterTypes,
  stringFilterTypes,
  tierFilterTypes,
} from '../types'
import { dateFilterToString } from './date/date_filter_to_string'
import { describeDate } from './date/describe_date'
import { describeLocation } from './location/describe_location'
import { locationToString } from './location/location_to_string'
import { describeNumber } from './number/describe_number'
import { numberToString } from './number/number_to_string'
import { describeString } from './string/describe_string'
import { stringFilterToString } from './string/string_filter_to_string'
import { describeTier } from './tier/describe_tier'
import { tierFilterToString } from './tier/tier_filter_to_string'
import { numberTransform } from './transform/numberTransform'
import { stringTransform } from './transform/stringTransform'

interface GrammarMapTypeOptions {
  grammar: string
  toString: FilterToStringFunctionType
  transform?: TransformFunction
  describe: FilterItemToStringFunction
  anyvalue: FilterASTNode
  subTypes: readonly string[]
}

type GrammarMapType = {
  [K in FilterExpressionType]: GrammarMapTypeOptions
}

/**
 * A map of available filter types and the grammar properties needed to parse and display
 */
const dateGrammarOptions: GrammarMapTypeOptions = {
  grammar: dateGrammar,
  toString: dateFilterToString,
  describe: describeDate,
  anyvalue: anyValue,
  subTypes: dateFilterTypes,
}

export const grammarsMap: GrammarMapType = {
  date: dateGrammarOptions,
  date_time: dateGrammarOptions,
  number: {
    grammar: numberGrammar,
    toString: numberToString,
    transform: numberTransform,
    describe: describeNumber,
    anyvalue: anyNumberItem,
    subTypes: numberFilterTypes,
  },
  string: {
    grammar: stringGrammar,
    toString: stringFilterToString,
    transform: stringTransform,
    describe: describeString,
    anyvalue: anyStringItem,
    subTypes: stringFilterTypes,
  },
  tier: {
    grammar: tierGrammar,
    toString: tierFilterToString,
    transform: stringTransform,
    describe: describeTier,
    anyvalue: anyStringItem,
    subTypes: tierFilterTypes,
  },
  location: {
    grammar: locationGrammar,
    toString: locationToString,
    describe: describeLocation,
    anyvalue: anyValue,
    subTypes: locationFilterTypes,
  },
}

export const typeToGrammar = (type: FilterExpressionType) =>
  grammarsMap[type] || {}
