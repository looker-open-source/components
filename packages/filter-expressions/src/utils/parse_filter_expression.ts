/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { Parser } from 'pegjs'
import { generate } from 'pegjs'
import type {
  FilterASTNode,
  FilterExpressionType,
  TransformFunction,
  UserAttributeWithValue,
} from '../types'
import { getMatchesAdvancedNode } from './get_matches_advanced_node'
import { transformAST } from './transform/transform_ast'
import { userAttributeTransform } from './transform/userAttributeTransform'
import { typeToGrammar } from './type_to_grammar'

/**
 * Generates a parser from a PEGjs grammar and caches the result
 */
const generateParser = (() => {
  const parserCache: { [key: string]: Parser } = {}
  return (type: string, grammar: string) => {
    if (!parserCache[type]) {
      parserCache[type] = generate(grammar)
    }
    return parserCache[type]
  }
})()

/**
 * A functions that uses a grammar of type type to parse an expression and returns an AST
 */
export const parseFilterExpression = (
  type: FilterExpressionType,
  expression: string,
  userAttributes?: UserAttributeWithValue[]
): FilterASTNode => {
  const {
    grammar,
    anyvalue,
    transform = (root: FilterASTNode) => root,
  } = typeToGrammar(type)
  if (expression === '') {
    return anyvalue
  }
  try {
    const parser = generateParser(type, grammar)
    const transforms: TransformFunction[] = [
      userAttributeTransform(userAttributes),
      transform,
    ]
    return transformAST(parser.parse(expression, { Object }), transforms)
  } catch (error) {
    return getMatchesAdvancedNode(expression)
  }
}
