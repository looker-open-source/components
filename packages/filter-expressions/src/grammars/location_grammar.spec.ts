/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { parseFilterExpression, summary } from '../utils'

const location = [
  ['36.97, -122.03', '36.97, -122.03'],
  ['-36.97, 122.03', '-36.97, 122.03'],
  ['-36.97, -122.03', '-36.97, -122.03'],
  ['40 miles from -36.97, -122.03', '40 miles from -36.97, -122.03'],
  ['40 miles from 36.97, -122.03', '40 miles from 36.97, -122.03'],
  ['100 miles from 36.97, -122.03', '100 miles from 36.97, -122.03'],
  [
    'inside box from 72.33, -173.14 to 14.39, -61.70',
    '72.3째N, 173.1째W to 14.4째N, 61.7째W',
  ],
  ['', 'is anywhere'],
  ['NOT NULL', 'is not null'],
  ['-NULL', 'is not null'],
  ['NULL', 'is null'],
]

describe('Location grammar can parse expressions', () => {
  it.each(location)('%s', (expression, result) => {
    expect(parseFilterExpression('location', expression)).toMatchSnapshot()
    expect(summary({ type: 'location', expression })).toBe(result)
  })
})
