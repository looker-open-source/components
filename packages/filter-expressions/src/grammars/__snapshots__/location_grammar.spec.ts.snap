// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location grammar can parse expressions  1`] = `
Object {
  "id": 1,
  "is": true,
  "type": "anyvalue",
}
`;

exports[`Location grammar can parse expressions -36.97, -122.03 1`] = `
Object {
  "id": 1,
  "lat": -36.97,
  "left": undefined,
  "lon": -122.03,
  "right": undefined,
  "type": "location",
}
`;

exports[`Location grammar can parse expressions -36.97, 122.03 1`] = `
Object {
  "id": 1,
  "lat": -36.97,
  "left": undefined,
  "lon": 122.03,
  "right": undefined,
  "type": "location",
}
`;

exports[`Location grammar can parse expressions -NULL 1`] = `
Object {
  "id": 1,
  "left": undefined,
  "right": undefined,
  "type": "notnull",
}
`;

exports[`Location grammar can parse expressions 36.97, -122.03 1`] = `
Object {
  "id": 1,
  "lat": 36.97,
  "left": undefined,
  "lon": -122.03,
  "right": undefined,
  "type": "location",
}
`;

exports[`Location grammar can parse expressions 40 miles from -36.97, -122.03 1`] = `
Object {
  "distance": 40,
  "id": 1,
  "lat": -36.97,
  "left": undefined,
  "lon": -122.03,
  "right": undefined,
  "type": "circle",
  "unit": "miles",
}
`;

exports[`Location grammar can parse expressions 40 miles from 36.97, -122.03 1`] = `
Object {
  "distance": 40,
  "id": 1,
  "lat": 36.97,
  "left": undefined,
  "lon": -122.03,
  "right": undefined,
  "type": "circle",
  "unit": "miles",
}
`;

exports[`Location grammar can parse expressions 100 miles from 36.97, -122.03 1`] = `
Object {
  "distance": 100,
  "id": 1,
  "lat": 36.97,
  "left": undefined,
  "lon": -122.03,
  "right": undefined,
  "type": "circle",
  "unit": "miles",
}
`;

exports[`Location grammar can parse expressions NOT NULL 1`] = `
Object {
  "id": 1,
  "left": undefined,
  "right": undefined,
  "type": "notnull",
}
`;

exports[`Location grammar can parse expressions NULL 1`] = `
Object {
  "id": 1,
  "left": undefined,
  "right": undefined,
  "type": "null",
}
`;

exports[`Location grammar can parse expressions inside box from 72.33, -173.14 to 14.39, -61.70 1`] = `
Object {
  "id": 1,
  "lat": 72.33,
  "lat1": 14.39,
  "left": undefined,
  "lon": -173.14,
  "lon1": -61.7,
  "right": undefined,
  "type": "box",
}
`;
